packages/ember-runtime/lib/mixins/array.js

```diff
@@ -141,19 +141,27 @@ export function arrayContentDidChange(array, startIdx, removeAmt, addAmt) {
 
   let meta = peekMeta(array);
   let cache = meta && meta.readableCache();
-
-  if (cache) {
-    if (cache.firstObject !== undefined &&
-        objectAt(array, 0) !== cacheFor.get(cache, 'firstObject')) {
-      propertyWillChange(array, 'firstObject', meta);
-      propertyDidChange(array, 'firstObject', meta);
-    }
-    if (cache.lastObject !== undefined &&
-        objectAt(array, get(array, 'length') - 1) !== cacheFor.get(cache, 'lastObject')) {
-      propertyWillChange(array, 'lastObject', meta);
-      propertyDidChange(array, 'lastObject', meta);
-    }
+  if (cache !== undefined) {
+    if (cache.firstObject !== undefined && startIdx === 0) {
+      propertyWillChange(array, 'firstObject');
+      propertyDidChange(array, 'firstObject');
+    }
+
+    if (cache.lastObject !== undefined) {
+      let addedAmount = (addAmt === -1 ? 0 : addAmt);
+      let removedAmount = (removeAmt === -1 ? 0 : removeAmt);
+      let delta = addedAmount - removedAmount;
+      let length = get(array, 'length');
+
+      let lastIndex = length - delta - 1;
+      let lastAffectedIndex = startIdx + removedAmount;
+      if (lastIndex < lastAffectedIndex) {
```

Comment:
I think this will miss cases where `lastObject` changes from negative indices, eg

```js
[0, 1, 2, 3].replace(-2, 2);
// => [0, 1]
```